### PR TITLE
Unit Tests and Bug Fixes

### PR DESCRIPTION
**Pull request by kevinazhu**
1. fixed bug with mismatching brackets
2. fixed bug with nested brackets not working
Also added some unit tests to test these features.

### CODE REVIEW

File: config_parser_test.cc [hunk 1]
```diff
@@ -9,3 +11,58 @@ TEST(NginxConfigParserTest, SimpleConfig) {
...
+    EXPECT_FALSE(ParseString("foo bar"));
+}
+
+TEST_F(NginxStringConfigTest, InvalidConfigBlock) {
+    EXPECT_FALSE(ParseString("server listen 80; }"));
+    EXPECT_FALSE(ParseString("server { listen 80;"));
+}
+
+TEST_F(NginxStringConfigTest, ValidConfigNested) {
+	EXPECT_TRUE(ParseString("server { listen {foo 80;}}"));
```
Comment (jteo1) [inline config_parser_test.cc]: Seems fine to me, but perhaps add some large sample config files as well instead of just testing some corner cases.

**Review by jteo1 — CHANGES_REQUESTED** [review]
Perhaps add some large sample config files instead of just testing some corner cases to make sure it works in larger settings.

File: config_parser.cc [hunk 2]
```diff
@@ -198,15 +199,19 @@ bool NginxConfigParser::Parse(std::istream* config_file, NginxConfig* config) {
...
       }
       config_stack.pop();
+      unmatched_brackets--;
     } else if (token_type == TOKEN_TYPE_EOF) {
-      if (last_token_type != TOKEN_TYPE_STATEMENT_END &&
-          last_token_type != TOKEN_TYPE_END_BLOCK) {
+      if ((last_token_type != TOKEN_TYPE_STATEMENT_END &&
+          last_token_type != TOKEN_TYPE_END_BLOCK) || unmatched_brackets != 0) {
         // Error.
         break;
```
Comment (jteo1) [inline config_parser.cc:216]: when it breaks here, you probably need to write some new output error message, because otherwise it defaults to something else

File: config_parser.cc [hunk 3]
```diff
@@ -198,18 +199,27 @@ bool NginxConfigParser::Parse(std::istream* config_file, NginxConfig* config) {
...
     } else if (token_type == TOKEN_TYPE_EOF) {
       if (last_token_type != TOKEN_TYPE_STATEMENT_END &&
           last_token_type != TOKEN_TYPE_END_BLOCK) {
         // Error.
         break;
       }
+      else if (unmatched_brackets != 0) {
+        // Error
+        printf ("Mismatched brackets\n");
+        return false;
```
Comment (dstephan1) [inline config_parser.cc:221]: Consider having a more descriptive error - what bracket is missing or how many are missing.
